File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/SubjectControllerImpl.java
package com.snowplowanalytics.snowplow.internal.tracker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.snowplowanalytics.snowplow.controller.SubjectController;
import com.snowplowanalytics.snowplow.internal.Controller;
import com.snowplowanalytics.snowplow.util.Size;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class SubjectControllerImpl extends Controller implements SubjectController {

    // Constructors

    public SubjectControllerImpl(@NonNull ServiceProviderInterface serviceProvider) {
        super(serviceProvider);
    }

    // Getters and Setters

    @Nullable
    @Override
    public String getUserId() {
        return getSubject().userId;
    }

    @Override
    public void setUserId(@Nullable String userId) {
        getDirtyConfig().userId = userId;
        getDirtyConfig().userIdUpdated = true;
        getSubject().setUserId(userId);
    }

    @Nullable
    @Override
    public String getNetworkUserId() {
        return getSubject().networkUserId;
    }

    @Override
    public void setNetworkUserId(@Nullable String networkUserId) {
        getDirtyConfig().networkUserId = networkUserId;
        getDirtyConfig().networkUserIdUpdated = true;
        getSubject().setNetworkUserId(networkUserId);
    }

    @Nullable
    @Override
    public String getDomainUserId() {
        return getSubject().domainUserId;
    }

    @Override
    public void setDomainUserId(@Nullable String domainUserId) {
        getDirtyConfig().domainUserId = domainUserId;
        getDirtyConfig().domainUserIdUpdated = true;
        getSubject().setDomainUserId(domainUserId);
    }

    @Nullable
    @Override
    public String getUseragent() {
        return getSubject().useragent;
    }

    @Override
    public void setUseragent(@Nullable String useragent) {
        getDirtyConfig().useragent = useragent;
        getDirtyConfig().useragentUpdated = true;
        getSubject().setUseragent(useragent);
    }

    @Nullable
    @Override
    public String getIpAddress() {
        return getSubject().ipAddress;
    }

    @Override
    public void setIpAddress(@Nullable String ipAddress) {
        getDirtyConfig().ipAddress = ipAddress;
        getDirtyConfig().ipAddressUpdated = true;
        getSubject().setIpAddress(ipAddress);
    }

    @Nullable
    @Override
    public String getTimezone() {
        return getSubject().timezone;
    }

    @Override
    public void setTimezone(@Nullable String timezone) {
        getDirtyConfig().timezone = timezone;
        getDirtyConfig().timezoneUpdated = true;
        getSubject().setTimezone(timezone);
    }

    @Nullable
    @Override
    public String getLanguage() {
        return getSubject().language;
    }

    @Override
    public void setLanguage(@Nullable String language) {
        getDirtyConfig().language = language;
        getDirtyConfig().languageUpdated = true;
        getSubject().setLanguage(language);
    }

    @Nullable
    @Override
    public Size getScreenResolution() {
        return getSubject().screenResolution;
    }

    @Override
    public void setScreenResolution(@Nullable Size screenResolution) {
        getDirtyConfig().screenResolution = screenResolution;
        getDirtyConfig().screenResolutionUpdated = true;
        getSubject().setScreenResolution(screenResolution.getWidth(), screenResolution.getHeight());
    }

    @Nullable
    @Override
    public Size getScreenViewPort() {
        return getSubject().screenViewPort;
    }

    @Override
    public void setScreenViewPort(@Nullable Size screenViewPort) {
        getDirtyConfig().screenViewPort = screenViewPort;
        getDirtyConfig().screenViewPortUpdated = true;
        getSubject().setViewPort(screenViewPort.getWidth(), screenViewPort.getHeight());
    }

    @Nullable
    @Override
    public Integer getColorDepth() {
        return getSubject().colorDepth;
    }

    @Override
    public void setColorDepth(@Nullable Integer colorDepth) {
        getDirtyConfig().colorDepth = colorDepth;
        getDirtyConfig().colorDepthUpdated = true;
        getSubject().setColorDepth(colorDepth);
    }

    // Private methods

    private Subject getSubject() {
        return serviceProvider.getOrMakeSubject();
    }

    private SubjectConfigurationUpdate getDirtyConfig() {
        return serviceProvider.getSubjectConfigurationUpdate();
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/SubjectControllerImpl.kt
/*
 * Copyright (c) 2015-present Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.tracker

import androidx.annotation.RestrictTo
import com.snowplowanalytics.core.Controller
import com.snowplowanalytics.snowplow.configuration.SubjectConfiguration
import com.snowplowanalytics.snowplow.controller.SubjectController
import com.snowplowanalytics.snowplow.util.Size

@RestrictTo(RestrictTo.Scope.LIBRARY)
class SubjectControllerImpl  // Constructors
    (serviceProvider: ServiceProviderInterface) : Controller(serviceProvider), SubjectController {
    // Getters and Setters
    override var userId: String?
        get() = subject.userId
        set(userId) {
            dirtyConfig.userId = userId
            subject.userId = userId
        }
    
    override var networkUserId: String?
        get() = subject.networkUserId
        set(networkUserId) {
            dirtyConfig.networkUserId = networkUserId
            subject.networkUserId = networkUserId
        }
    
    override var domainUserId: String?
        get() = subject.domainUserId
        set(domainUserId) {
            dirtyConfig.domainUserId = domainUserId
            subject.domainUserId = domainUserId
        }
    
    override var useragent: String?
        get() = subject.useragent
        set(useragent) {
            dirtyConfig.useragent = useragent
            subject.useragent = useragent
        }
    
    override var ipAddress: String?
        get() = subject.ipAddress
        set(ipAddress) {
            dirtyConfig.ipAddress = ipAddress
            subject.ipAddress = ipAddress
        }
    
    override var timezone: String?
        get() = subject.timezone
        set(timezone) {
            dirtyConfig.timezone = timezone
            subject.timezone = timezone
        }
    
    override var language: String?
        get() = subject.language
        set(language) {
            dirtyConfig.language = language
            subject.language = language
        }
    
    override var screenResolution: Size?
        get() = subject.screenResolution
        set(screenResolution) {
            dirtyConfig.screenResolution = screenResolution
            subject.screenResolution = screenResolution
        }
    
    override var screenViewPort: Size?
        get() = subject.screenViewPort
        set(screenViewPort) {
            dirtyConfig.screenViewPort = screenViewPort
            subject.screenViewPort = screenViewPort
        }
    
    override var colorDepth: Int?
        get() = subject.colorDepth
        set(colorDepth) {
            dirtyConfig.colorDepth = colorDepth
            subject.colorDepth = colorDepth
        }

    // Private methods
    private val subject: Subject
        get() = serviceProvider.getOrMakeSubject()
    private val dirtyConfig: SubjectConfiguration
        get() = serviceProvider.subjectConfiguration
}
